src/main/java/org/kuali/kra/budget/document/BudgetDocumentRule.java

```diff
@@ -154,7 +154,86 @@ protected boolean processCustomSaveDocumentBusinessRules(Document document) {
     * @param budgetDocument
     * @return
     */
-    @SuppressWarnings("rawtypes")
+    @SuppressWarnings("rawtypes")  
+	protected boolean processBudgetProjectIncomeBusinessRule(BudgetDocument budgetDocument) {
+        boolean valid = true;
+        MessageMap errorMap = GlobalVariables.getMessageMap();
+        int i = 0;
+        
+        // Get each budget cost share.
+        for (BudgetCostShare budgetCostShare : budgetDocument.getBudget().getBudgetCostShares()) {
+            String errorPath = "budgetCostShare[" + i + "]";
+            errorMap.addToErrorPath(errorPath);
+            
+            // Validate share percentage.
+            if(budgetCostShare.getSharePercentage() != null && (budgetCostShare.getSharePercentage().isLessThan(new BudgetDecimal(0)) || 
+               budgetCostShare.getSharePercentage().isGreaterThan(new BudgetDecimal(100)))) {
+                errorMap.putError("sharePercentage", KeyConstants.ERROR_COST_SHARE_PERCENTAGE);
+                valid = false;
+            }
+            
+            // Check for duplicate fiscal year, source account and unit number on all unchecked cost shares.
+            if (i < budgetDocument.getBudget().getBudgetCostShareCount()) {
+                for (int j = i + 1; j < budgetDocument.getBudget().getBudgetCostShareCount(); j++) {
+                    BudgetCostShare tmpCostShare = budgetDocument.getBudget().getBudgetCostShare(j);
+                    int thisFiscalYear = budgetCostShare.getProjectPeriod() == null ? Integer.MIN_VALUE : budgetCostShare.getProjectPeriod();
+                    int otherFiscalYear = tmpCostShare.getProjectPeriod() == null ? Integer.MIN_VALUE : tmpCostShare.getProjectPeriod();
+                    
+                    //Only validate for same fiscal year
+                    if (thisFiscalYear == otherFiscalYear) {                        
+					    String thisSourceAccount = budgetCostShare.getSourceAccount() == null ? "" : budgetCostShare.getSourceAccount().trim();
+					    String otherSourceAccount = tmpCostShare.getSourceAccount() == null ? "" : tmpCostShare.getSourceAccount().trim();                   
+					    String thisSourceUnitNumber = budgetCostShare.getSourceUnitNumber() == null ? "" : budgetCostShare.getSourceUnitNumber().trim();
+					    String otherSourceUnitNumber = tmpCostShare.getSourceUnitNumber() == null ? "" : tmpCostShare.getSourceUnitNumber().trim();					    
```

Comment:
Consider creating a private helper to clear out all of these ternaries for easier reading, both here and further down.